Source/ASRunLoopQueue.mm

```diff
@@ -469,3 +489,230 @@ - (void)unlock
+
+// CoreAnimation commit order is 2000000, the goal of this is to process shortly beforehand
+// but after most other scheduled work on the runloop has processed.
+static int const kASASCATransactionQueueOrder = 1000000;
+
++ (ASCATransactionQueue *)sharedQueue
+{
+  static dispatch_once_t onceToken;
+  static ASCATransactionQueue *sharedQueue;
+  dispatch_once(&onceToken, ^{
+    sharedQueue = [[ASCATransactionQueue alloc] init];
+  });
+  return sharedQueue;
+}
+
+- (instancetype)init
+{
+  if (self = [super init]) {
+    _runLoop = CFRunLoopGetMain();
+    NSPointerFunctionsOptions options = NSPointerFunctionsStrongMemory;
+    _internalQueue = [[NSPointerArray alloc] initWithOptions:options];
+
+    // We don't want to pollute the top-level app activities with run loop batches, so we create one top-level
+    // activity per queue, and each batch activity joins that one instead.
+    _rootActivity = as_activity_create("Process run loop queue items", OS_ACTIVITY_NONE, OS_ACTIVITY_FLAG_DEFAULT);
+    {
+      // Log a message identifying this queue into the queue's root activity.
+      as_activity_scope_verbose(_rootActivity);
+      as_log_verbose(ASDisplayLog(), "Created run loop queue: %@", self);
+    }
+
+    // Self is guaranteed to outlive the observer.  Without the high cost of a weak pointer,
+    // __unsafe_unretained allows us to avoid flagging the memory cycle detector.
+    __unsafe_unretained __typeof__(self) weakSelf = self;
+    void (^handlerBlock) (CFRunLoopObserverRef observer, CFRunLoopActivity activity) = ^(CFRunLoopObserverRef observer, CFRunLoopActivity activity) {
+      [weakSelf processQueue];
+    };
+    _runLoopObserver = CFRunLoopObserverCreateWithHandler(NULL, kCFRunLoopBeforeWaiting, true, kASASCATransactionQueueOrder, handlerBlock);
+    CFRunLoopAddObserver(_runLoop, _runLoopObserver,  kCFRunLoopCommonModes);
+
+    // It is not guaranteed that the runloop will turn if it has no scheduled work, and this causes processing of
+    // the queue to stop. Attaching a custom loop source to the run loop and signal it if new work needs to be done
+    CFRunLoopSourceContext sourceContext = {};
+    sourceContext.perform = runLoopSourceCallback;
+#if ASRunLoopQueueLoggingEnabled
+    sourceContext.info = (__bridge void *)self;
+#endif
+    _runLoopSource = CFRunLoopSourceCreate(NULL, 0, &sourceContext);
+    CFRunLoopAddSource(_runLoop, _runLoopSource, kCFRunLoopCommonModes);
+
+#if ASRunLoopQueueLoggingEnabled
+    _runloopQueueLoggingTimer = [NSTimer timerWithTimeInterval:1.0 target:self selector:@selector(checkRunLoop) userInfo:nil repeats:YES];
+    [[NSRunLoop mainRunLoop] addTimer:_runloopQueueLoggingTimer forMode:NSRunLoopCommonModes];
+#endif
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  if (CFRunLoopContainsSource(_runLoop, _runLoopSource, kCFRunLoopCommonModes)) {
+    CFRunLoopRemoveSource(_runLoop, _runLoopSource, kCFRunLoopCommonModes);
+  }
+  CFRelease(_runLoopSource);
```

Comment:
We can unconditionally remove the source – documentation says that if the source isn't present the function does nothing.